src/util/dataAccess.cpp

```diff
@@ -165,21 +165,114 @@ ndsize_t positionToIndex(double position, const string &unit, const RangeDimensi
+    } else if (dt == DimensionType::Set) {
+        SetDimension sd = dim.asSetDimension();
+        pos = 0.0;
+        if (max_index > pow(FLT_RADIX, std::numeric_limits<double>::digits)) {
+            throw nix::OutOfBounds("dataAccess::fillPositionsExtents: shape cannot be cast to double without loss of precision. Please open an issue on github!");
+        }
+        ext = static_cast<double>(max_index);
+    }
+}
+
+
+vector<pair<double, double>> maximumExtents(const DataArray &array) {
+    vector<Dimension> dimensions = array.dimensions();
+    vector<pair<double, double>> max_extents;
+    for (size_t i = 0; i < dimensions.size(); ++i) {
+        double pos, ext;
+        getMaxExtent(dimensions[i], array.dataExtent()[i]-1, pos, ext);
+        max_extents.emplace_back(pos, ext);
+    }
+    return max_extents;
+}
+
+
+string getDimensionUnit(const Dimension &dim) {
+    DimensionType dim_type = dim.dimensionType();
+    string unit;
+    switch (dim_type) {
+    case DimensionType::Set:
+        unit = "none";
+        break;
+    case DimensionType::DataFrame: {
+        DataFrameDimension df_dim = dim.asDataFrameDimension();
+        unit = "none";
+        if (df_dim.columnIndex()) {
+            unit = df_dim.unit();
+        }
+        if (unit.empty()) {
+            unit = "none";
+        }
+        break;
+    }
+    case DimensionType::Sample: {
+        SampledDimension sd = dim.asSampledDimension();
+        unit = sd.unit() ? *sd.unit() : "none";
+        break;
+    }
+    case DimensionType::Range: {
+        RangeDimension rd = dim.asRangeDimension();
+        unit = rd.unit() ? *rd.unit() : "none";
+        break;
+    }
+    default:
+       throw std::invalid_argument("Unkown DimensionType");
+    }
+    return unit;
+}
+
+
 void getOffsetAndCount(const Tag &tag, const DataArray &array, NDSize &offset, NDSize &count) {
     vector<double> position = tag.position();
     vector<double> extent = tag.extent();
     vector<string> units = tag.units();
-    NDSize temp_offset(position.size());
-    NDSize temp_count(position.size(), 1);
     vector<Dimension> dimensions = array.dimensions();
+    size_t dim_count = dimensions.size();
+    NDSize shape = array.dataExtent();
+
+    if (extent.size() > 0 && (extent.size() != position.size())) {
+        throw IncompatibleDimensions("Size of position and extent do not match!",
+                                     "nix::util::getOffsetAndCount");
+    }
+    vector<pair<double, double>> max_extents;
+    if (position.size() < dim_count) {
+        max_extents = maximumExtents(array);
+    }
+    if (extent.size() == 0) {
+        extent.resize(position.size(), 0.0);
+    }
+    while (position.size() > dim_count) {
+        position.pop_back();
+        extent.pop_back();
+    }
+    while (position.size() < dim_count) {
+        position.push_back(get<0>(max_extents[position.size()]));
+        extent.push_back(get<1>(max_extents[extent.size()]));
```

Comment:
I am curious if the compiler is smart enough to optimize the access of the pair.

Reply:
you got a suggestion?

Reply:
Nothing obvious, I think.